### ISSUE TITLE
Waiting for completion does not work when minions are not connected

### ISSUE THREAD

**Issue by blaketmiller** [jenkinsci/saltstack-plugin#38]
Greetings friend! Interesting observation I made a moment ago. Doing a simple ping across all my minions when one of the minions is disconnected will make the job run forever. In this example, `cache1` is disconnected:

```
saltmaster$ salt '*' test.ping
web1:
    True
web2:
    True
cache1:
    Minion did not return. [Not connected]
```

and the API job call will look something like:

```
Waiting for 3 minions
2 minions are done
Will check status every 10 seconds...
```

for eternity, even if `cache1` is reconnected after a period of time. I could increase the poll time perhaps. Is there something that the Salt API returns that can be checked to say "wait for completion", but only for connected minions? Feel free to toss this in the "not a bug, it's a feature" category if you like, just wanted to gather a thought on this 👍

**Comment by mchugh19** [comment]
Hey! Can you send over the debug output? That should include the response from the saltapi. 

It is likely that it only responds with the currently active minions and gives us no indication that any are unreachable. As a work around, we could implement a timer that kicks off when any minion returns. Let's say from the first minion coming back, the rest have 30 seconds to respond, or jenkins fails the job.

**Comment by blaketmiller** [comment]
Hmm, so all that logs show is:
```
Sending JSON: [{"client":"local","tgt":"lx12345","expr_form":"glob","fun":"test.ping"}]
```

but here's the potentially useful event bus log from salt master:
```
salt/job/20160916141543931800/new	{
    "_stamp": "2016-09-16T21:15:43.932613", 
    "arg": [], 
    "fun": "test.ping", 
    "jid": "20160916141543931800", 
    "minions": [
        "lx12345"
    ], 
    "tgt": "lx12345", 
    "tgt_type": "glob", 
    "user": "blmiller"
}
salt/run/20160916141543972910/new	{
    "_stamp": "2016-09-16T21:15:47.051087", 
    "fun": "runner.jobs.lookup_jid", 
    "jid": "20160916141543972910", 
    "user": "blmiller"
}
salt/run/20160916141543972910/ret	{
    "_stamp": "2016-09-16T21:15:53.175030", 
    "fun": "runner.jobs.lookup_jid", 
    "jid": "20160916141543972910", 
    "return": {}, 
    "success": true, 
    "user": "blmiller"
}
salt/run/20160916141553206287/new	{
    "_stamp": "2016-09-16T21:15:56.360101", 
    "fun": "runner.jobs.list_job", 
    "jid": "20160916141553206287", 
    "user": "blmiller"
}
salt/run/20160916141553206287/ret	{
    "_stamp": "2016-09-16T21:15:59.420905", 
    "fun": "runner.jobs.list_job", 
    "jid": "20160916141553206287", 
    "return": {
        "Arguments": [], 
        "Function": "test.ping", 
        "Minions": [
            "lx12345"
        ], 
        "Result": {}, 
        "StartTime": "2016, Sep 16 14:15:43.931800", 
        "Target": "lx12345", 
        "Target-type": "glob", 
        "User": "blmiller", 
        "jid": "20160916141543931800"
    }, 
    "success": true, 
    "user": "blmiller"
}
salt/run/20160916141609460582/new	{
    "_stamp": "2016-09-16T21:16:12.517203", 
    "fun": "runner.jobs.lookup_jid", 
    "jid": "20160916141609460582", 
    "user": "blmiller"
}
salt/run/20160916141609460582/ret	{
    "_stamp": "2016-09-16T21:16:18.676451", 
    "fun": "runner.jobs.lookup_jid", 
    "jid": "20160916141609460582", 
    "return": {}, 
    "success": true, 
    "user": "blmiller"
}
salt/run/20160916141618706246/new	{
    "_stamp": "2016-09-16T21:16:21.757416", 
    "fun": "runner.jobs.list_job", 
    "jid": "20160916141618706246", 
    "user": "blmiller"
}
salt/run/20160916141618706246/ret	{
    "_stamp": "2016-09-16T21:16:24.825387", 
    "fun": "runner.jobs.list_job", 
    "jid": "20160916141618706246", 
    "return": {
        "Arguments": [], 
        "Function": "test.ping", 
        "Minions": [
            "lx12345"
        ], 
        "Result": {}, 
        "StartTime": "2016, Sep 16 14:15:43.931800", 
        "Target": "lx12345", 
        "Target-type": "glob", 
        "User": "blmiller", 
        "jid": "20160916141543931800"
    }, 
    "success": true, 
    "user": "blmiller"
}
salt/run/20160916141634862856/new	{
    "_stamp": "2016-09-16T21:16:37.955097", 
    "fun": "runner.jobs.lookup_jid", 
    "jid": "20160916141634862856", 
    "user": "blmiller"
}
salt/run/20160916141634862856/ret	{
    "_stamp": "2016-09-16T21:16:44.108713", 
    "fun": "runner.jobs.lookup_jid", 
    "jid": "20160916141634862856", 
    "return": {}, 
    "success": true, 
    "user": "blmiller"
}
salt/run/20160916141644138943/new	{
    "_stamp": "2016-09-16T21:16:47.237960", 
    "fun": "runner.jobs.list_job", 
    "jid": "20160916141644138943", 
    "user": "blmiller"
}
salt/run/20160916141644138943/ret	{
    "_stamp": "2016-09-16T21:16:50.299865", 
    "fun": "runner.jobs.list_job", 
    "jid": "20160916141644138943", 
    "return": {
        "Arguments": [], 
        "Function": "test.ping", 
        "Minions": [
            "lx12345"
        ], 
        "Result": {}, 
        "StartTime": "2016, Sep 16 14:15:43.931800", 
        "Target": "lx12345", 
        "Target-type": "glob", 
        "User": "blmiller", 
        "jid": "20160916141543931800"
    }, 
    "success": true, 
    "user": "blmiller"
}
```

which I think is illustrating exactly what you described and I think your proposed solution is probably the best/only way to "fix" it.

Is that timer a feature request that fits within the scope of your plugin or should this be directed toward Salt itself?

**Comment by mchugh19** [comment]
Totally in scope. Turns out that's exactly how salt cli does it as well. So
just need a fee minutes to work it up. Otherwise PRs accepted ;)

**Comment by mchugh19** [comment]
Merged into master. Will be in next release.

**Comment by mchugh19** [comment]
Released 2.0
